contractsV5/eth2dai/Eth2DaiReserve.sol

```diff
@@ -0,0 +1,846 @@
+    }
+
+    function takeMatchingOrders(ERC20 destToken, uint srcAmount, OfferData[] memory offers)
+        internal
+        returns(uint actualDestAmount)
+    {
+        require(destToken != ETH_TOKEN_ADDRESS, "takeMatchingOrders: destToken is ETH");
+
+        uint lastReserveBalance = destToken.balanceOf(address(this));
+        uint remainingSrcAmount = srcAmount;
+
+        for (uint i = 0; i < offers.length; i++) {
+            if (offers[i].id == 0 || remainingSrcAmount == 0) { break; }
+
+            uint payAmount = minOf(remainingSrcAmount, offers[i].payAmount);
+            uint buyAmount = payAmount * offers[i].buyAmount / offers[i].payAmount;
+
+            otc.take(bytes32(offers[i].id), uint128(buyAmount));
+            remainingSrcAmount -= payAmount;
+        }
+
+        // must use all amount
+        require(remainingSrcAmount == 0, "takeMatchingOrders: did not take all src amount");
+
+        uint newReserveBalance = destToken.balanceOf(address(this));
+
+        require(newReserveBalance > lastReserveBalance, "takeMatchingOrders: newReserveBalance <= lastReserveBalance");
+
+        actualDestAmount = newReserveBalance - lastReserveBalance;
+    }
+
+    function shouldUseInternalInventory(
+        ERC20 token,
+        uint tokenVal,
+        uint ethVal,
+        bool ethToToken,
+        OfferData memory bid,
+        OfferData memory ask
+    )
+        internal view
+        returns(bool shouldUse, uint premiumBps)
+    {
+        shouldUse = false;
+        premiumBps = 0;
+
+        if (!isInternalInventoryEnabled[address(token)]) { return (shouldUse, premiumBps); }
+
+        if (tokenVal <= MAX_QTY) { return (shouldUse, premiumBps); }
+
+        InternalInventoryData memory inventoryData = getInternalInventoryData(token);
+        premiumBps = inventoryData.premiumBps;
+
+        uint tokenBalance = token.balanceOf(address(this));
+
+        if (ethToToken) {
+            if (tokenBalance < tokenVal) { return (shouldUse, premiumBps); }
+            if (tokenBalance - tokenVal < inventoryData.minTokenBal) { return (shouldUse, premiumBps); }
+        } else {
+            if (address(this).balance < ethVal) { return (shouldUse, premiumBps); }
+            if (tokenBalance + tokenVal > inventoryData.maxTokenBal) { return (shouldUse, premiumBps); }
+        }
+
+        if (!checkValidSpread(bid, ask, true, inventoryData.minSpreadBps)) {
+            return (shouldUse, premiumBps);
+        }
+
+        shouldUse = true;
+    }
+
+    function applyInternalInventoryHintToRate(
+        uint rate,
+        bool useInternalInventory
+    )
+        internal pure
+        returns(uint)
+    {
+        return rate % 2 == (useInternalInventory ? 1 : 0)
+            ? rate
+            : rate - 1;
+    }
+
+    function valueAfterReducingFee(uint val) internal view returns(uint) {
+        require(val <= MAX_QTY, "valueAfterReducingFee: val > MAX_QTY");
+        return ((10000 - feeBps) * val) / 10000;
```

Comment:
use BPS

Reply:
we have 2 values here:
BPS: 10000 (10^4) for this logics
BPS: 100000 (10^5) for token factor data
How should we name them?